### DISCUSSION TITLE
Deuterium in SMILES leads to invalid canonical tautomer

### DISCUSSION THREAD

**Discussion by kienerj** [rdkit/rdkit#5348] [General]
**Describe the bug**

A molecule with several deuterium atoms when converted to the canonical tautomer and then export as SMILES leads to invalid SMILES (5-valent carbon).

This issue is related to #4783

As in above case the issue happen inside a pipeline.

**To Reproduce**

```
mx = Chem.MolFromSmiles("[2H]C([2H])(C(=O)O)C([2H])([2H])C=O")
mt = MolStandardize.rdMolStandardize.CanonicalTautomer(mx)
Chem.MolFromSmiles(Chem.MolToSmiles(mt))
```
(no output as no molecule can be generated from the smiles)

I also looked at issue #4783 and tried it that way:

```
tautomer_params = Chem.MolStandardize.rdMolStandardize.CleanupParameters()
tautomer_params.tautomerRemoveIsotopicHs = True
enumerator = Chem.MolStandardize.rdMolStandardize.TautomerEnumerator(tautomer_params)
mol_can = Chem.RemoveHs(Chem.AddHs(enumerator.Canonicalize(mx)))
```

```
---------------------------------------------------------------------------
AtomValenceException                      Traceback (most recent call last)
C:\Temp/ipykernel_25680/2485090042.py in <module>
      2 tautomer_params.tautomerRemoveIsotopicHs = True
      3 enumerator = Chem.MolStandardize.rdMolStandardize.TautomerEnumerator(tautomer_params)
----> 4 mol_can = Chem.RemoveHs(Chem.AddHs(enumerator.Canonicalize(mx)))

AtomValenceException: Explicit valence for atom # 6 C, 5, is greater than permitted

```

**Expected behavior**

Get a valid molecule as output.

**Configuration (please complete the following information):**
 - RDKit version:  2022.03.1
 - OS: [e.g. Ubuntu 20.04] Win 10
 - Python version (if relevant): 3.9.7
 - Are you using conda? yes
 - If you are using conda, which channel did you install the rdkit from? conda-forge

**Comment by ptosco** [comment]
@kienerj This is not actually a bug. Let me try and explain below.
The `TautomerEnumerator` uses a set of SMARTS that will only work if the molecule has _only_ implicit Hs - it will fail in the presence of explicit Hs. For instance, with your starting molecule it will add an extra implicit H (evidenced in boldface in the snippet below) on a carbon that already has all its valences filled by explicit Hs:
<pre>
mx = Chem.MolFromSmiles("[2H]C([2H])(C(=O)O)C([2H])([2H])C=O")
mt = rdMolStandardize.CanonicalTautomer(mx)
Chem.MolToSmiles(mt)
'[2H]C([2H])(C(=O)O)<b>[CH]</b>([2H])([2H])C=O'
</pre>

Note that this behavior does not depend on the isotopic nature of Hs - the same will happen with 1H:
<pre>
smiles_params = Chem.SmilesParserParams()
smiles_params.removeHs = False
mx = Chem.MolFromSmiles("[H]C([H])(C(=O)O)C([H])([H])C=O", smiles_params)
mt = rdMolStandardize.CanonicalTautomer(mx)
Chem.MolToSmiles(mt)
'[H]C([H])(C(=O)O)<b>[CH]</b>([H])([H])C=O'
</pre>
To avoid this issue I have implemented the isotope tracking feature.
With an appropriate parameter, you can remove explicit isotopic Hs and track them through a private property on the parent heavy atom:
```
mx = Chem.MolFromSmiles("[2H]C([2H])(C(=O)O)C([2H])([2H])C=O")
remove_hs_params = Chem.RemoveHsParameters()
remove_hs_params.removeAndTrackIsotopes = True
mx_no_h = Chem.RemoveHs(mx, remove_hs_params)
mx_no_h
```
![image](https://user-images.githubusercontent.com/5244385/171275190-446f0004-3777-4b54-9999-659b659abee8.png)
```
for a in mx_no_h.GetAtoms():
    print(a.GetSymbol(), list(a.GetPropNames(includeComputed=True, includePrivate=True)))
C ['__computedProps', '_isotopicHs']
C ['__computedProps']
O ['__computedProps']
O ['__computedProps']
C ['__computedProps', '_isotopicHs']
C ['__computedProps']
O ['__computedProps']
```
Now you can enumerate tautomers. By default, any isotopic H involved in a tautomeric rearrangement is removed, in the assumption that it will be exchanged with solvent. We know that many rearrangement attempted by the tautomer enumerator actually do not take place at room temperature in normal conditions, so there is a `tautomerRemoveIsotopicHs` option (that defaults to `True`) that can be set to `False` to retain isotopic Hs also when they are involved in tautomeric rearrangement:
```
tautomer_params = rdMolStandardize.CleanupParameters()
tautomer_params.tautomerRemoveIsotopicHs = False
mt = rdMolStandardize.CanonicalTautomer(mx_no_h, tautomer_params)
mt
```
![image](https://user-images.githubusercontent.com/5244385/171276216-d25a58f0-558e-4abd-beb6-16399ba45610.png)
As expected, the canonical tautomer is identical to the input molecule. Since we set `tautomer_params.tautomerRemoveIsotopicHs` to `False`, deuterium atom positions should have been preserved.
Let's restore them by adding all Hs (including isotopic ones), and then removing only non isotopic Hs:
```
mt_isotopic_h = Chem.RemoveHs(Chem.AddHs(mt))
mt_isotopic_h
```
![image](https://user-images.githubusercontent.com/5244385/171276509-cd032512-9bd5-4c91-9aa9-21a9b6137794.png)
As expected, the deuterium atoms were restored.
As I believe the above is not obvious, if you agree I will move this to Discussions so it may help others.

Please let me know if you agree.

**Comment by kienerj** [comment]
Ok I understand.

But it means that when automatically processing a large number of molecules, all molecules need to go through this process.
One could check if the molecule even has deuterium and only if do this processing. 

What happens with "chiral" H atoms? Do they need to be removed to and how would the information then be preserved?

**Comment by ptosco** [comment]
@kienerj Not sure what you mean by "chiral" H atoms. If you mean something like L-aminoacids, by default chirality at centers involved in tautomeric rearrangements will be removed as it is assumed they will epimerize. So:
```
mx = Chem.MolFromSmiles("C[C@@H](C(=O)O)N")
mx
```
![image](https://user-images.githubusercontent.com/5244385/171347581-07dce039-375d-4d9c-b4ee-e7a305261084.png)
```
mt = rdMolStandardize.CanonicalTautomer(mx)
mt
```
![image](https://user-images.githubusercontent.com/5244385/171347652-9f58531a-ce9e-433f-b67c-0d0a05eac35b.png)
There is a flag to change this default:
```
tautomer_params = rdMolStandardize.CleanupParameters()
tautomer_params.tautomerRemoveSp3Stereo = False
mt = rdMolStandardize.CanonicalTautomer(mx, tautomer_params)
mt
```
![image](https://user-images.githubusercontent.com/5244385/171347719-e2ec3676-61bd-4c01-940d-d457ea784fc6.png)
Similarly, there is a `tautomerRemoveBondStereo` flag for double bond stereochemistry.

If you were meaning something different, could you please expand a bit?

Also, you have not replied to my question if I can move this thread to the Discussion tab.
If you do not explicitly disagree in your next reply, I will move it.

**Comment by kienerj** [comment]
Hi Paolo, yes please move to discussion tab.

About chiral H I meant something like below (albeit that is just a quickly made example which makes limited sense)

![image](https://user-images.githubusercontent.com/1567309/171366866-0cf31ed6-c013-4d91-bd8a-35efac19dac0.png)

I was wondering if the could also pose issues for the tautomer enumeration if not removed beforehand?

**Comment by ptosco** [comment]
@kienerj If such a molecule was drawn in a sketcher it will translate to something like:
```

  MJ201100                      

 12 13  0  0  0  0  0  0  0  0999 V2000
   -2.9017    1.2491    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -3.6162    0.8366    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -3.6162    0.0115    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.9017   -0.4008    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.1872    0.0115    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.1872    0.8366    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.4727   -0.4009    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.7583    0.0115    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.7582    0.8365    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.4728    1.2491    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.1872    1.6616    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
   -2.1873   -0.8134    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  1  6  1  0  0  0  0
  2  3  1  0  0  0  0
  3  4  1  0  0  0  0
  4  5  1  0  0  0  0
  5  6  1  0  0  0  0
  7  8  1  0  0  0  0
  8  9  1  0  0  0  0
  9 10  1  0  0  0  0
  5  7  1  0  0  0  0
  6 10  1  0  0  0  0
  6 11  1  1  0  0  0
  5 12  1  6  0  0  0
M  END
```
or `[H][C@@]12CCCC[C@@]1([H])CCCC2` in SMILES notation.
In both cases, reading the molblock or the SMILES in RDKIt by default will remove all explicit Hs and convert them to implicit while preserving stereochemistry:
```
mx = Chem.MolFromMolBlock("""
  MJ201100                      

 12 13  0  0  0  0  0  0  0  0999 V2000
   -2.9017    1.2491    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -3.6162    0.8366    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -3.6162    0.0115    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.9017   -0.4008    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.1872    0.0115    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.1872    0.8366    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.4727   -0.4009    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.7583    0.0115    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.7582    0.8365    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.4728    1.2491    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.1872    1.6616    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
   -2.1873   -0.8134    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  1  6  1  0  0  0  0
  2  3  1  0  0  0  0
  3  4  1  0  0  0  0
  4  5  1  0  0  0  0
  5  6  1  0  0  0  0
  7  8  1  0  0  0  0
  8  9  1  0  0  0  0
  9 10  1  0  0  0  0
  5  7  1  0  0  0  0
  6 10  1  0  0  0  0
  6 11  1  1  0  0  0
  5 12  1  6  0  0  0
M  END
""")
mx
```
![image](https://user-images.githubusercontent.com/5244385/171374222-4942b8dd-d44c-4cbe-94ae-d4ca5ed502ec.png)

From the drawing it seems that explicit Hs are still there but actually they are not in the molecule graph anymore; RDKit adds them only to the 2D drawing to make the stereochemistry clearer, but the wedged bonds are actually located between carbon atoms as you can see printing the molblock after reading it into RDKit:
```
print(Chem.MolToMolBlock(mx))

     RDKit          2D

 10 11  0  0  0  0  0  0  0  0999 V2000
   -2.9017    1.2491    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -3.6162    0.8366    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -3.6162    0.0115    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.9017   -0.4008    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.1872    0.0115    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.1872    0.8366    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.4727   -0.4009    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.7583    0.0115    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.7582    0.8365    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.4728    1.2491    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0
  6  1  1  6
  2  3  1  0
  3  4  1  0
  5  4  1  1
  5  6  1  0
  7  8  1  0
  8  9  1  0
  9 10  1  0
  5  7  1  0
  6 10  1  0
M  END
```
Since there are no explicit Hs in the molecule graph, the tautomer canonicalization won't incur into problems.
If you set `tautomerRemoveSp3Stereo` to `False`, stereochemistry will also be preserved during tautomer canonicalization:
```
tautomer_params = rdMolStandardize.CleanupParameters()
tautomer_params.tautomerRemoveSp3Stereo = False
mt = rdMolStandardize.CanonicalTautomer(mx, tautomer_params)
mt
```
![image](https://user-images.githubusercontent.com/5244385/171374653-b1764093-aaf1-4fb7-96f2-8212561dd774.png)
I hope this answers your question.

**Reply by kienerj** [reply]
Many thanks Paolo for these extensive explanations.

I have another question regarding TautomerQuery. I assume this has the same requirement regarding explicit Hs? 

If a user draws explicit Hs to "block a position" for a substructure search and how to then handle that if the search needs to be tautomer-aware? AdjustQueryParameters can be to strict if say the user want to block only 1 specific position in a benzene ring but allow all others to change (have a different degree compared to query).

**Comment by ptosco** [comment]
@kienerj Hi Joos, you can use `MergeQueryHs()` to convert the explicit Hs sketched in MarvinJS (or another sketcher, or entered as SMILES as in my example below) into implicit query Hs:
```
from rdkit import Chem

mol1 = Chem.MolFromSmiles("c1cc(C)nc(C)c1")
mol2 = Chem.MolFromSmiles("c1ccncc1")

params = Chem.SmilesParserParams()
params.removeHs = False
query = Chem.MolFromSmiles("c1ccnc([H])c1", params)
query_merged = Chem.MergeQueryHs(query)

mol1.HasSubstructMatch(query_merged)
False

mol2.HasSubstructMatch(query_merged)
True
```
